src/main/java/com/machowina/service/TicketServiceImp.java

```diff
@@ -0,0 +1,181 @@
+package com.machowina.service;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.List;
+
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.machowina.exception.EntityNotFoundException;
+import com.machowina.exception.TicketAlreadyRunning;
+import com.machowina.exception.TicketAlreadyStopped;
+import com.machowina.exception.TicketNotStoppedException;
+import com.machowina.model.Car;
+import com.machowina.model.ParkingRates;
+import com.machowina.model.ParkingTicket;
+import com.machowina.model.ParkingZone;
+import com.machowina.model.User;
+import com.machowina.repository.TicketRepository;
+
+@Service
+public class TicketServiceImp implements TicketService {
+	
+	private final CarService carService;
+	private final ParkingZoneService zoneService;
+	private final ParkingRatesService parkingRatesService;
+	
+	private final TicketRepository ticketRepository;
+	
+
+	public TicketServiceImp(CarService carService, ParkingZoneService zoneService,
+			TicketRepository ticketRepository, ParkingRatesService parkingRatesService) {
+		super();
+		this.carService = carService;
+		this.zoneService = zoneService;
+		this.ticketRepository = ticketRepository;
+		this.parkingRatesService = parkingRatesService;
+	}
+
+	@Override
+	@Transactional
+	public Long generateNewTicket(Long carId, Long zoneId) {
+		
+		checkForDuplicatingTicket(carId);
+		
+		ParkingTicket newTicket = createTicket(carId, zoneId);
+		
+		return saveTicket(newTicket);
+	
+	}
+	
+
+	@Override
+	public ParkingTicket createTicket(Long carId, Long zoneId) {
+		
+		Car car = carService.findById(carId);
+		User driver = car.getDriver();
+		ParkingZone zone = zoneService.findOne(zoneId);
+		LocalDateTime startTime = LocalDateTime.now();
+		
+		ParkingTicket newTicket = new ParkingTicket(startTime, zone, car, driver);
+		
+		return newTicket;
+		
+	}
+
+	@Override
+	public void checkForDuplicatingTicket(Long carId) {
+		ParkingTicket ticket = ticketRepository.findOneByCarIdAndIsStoppedFalse(carId);
+		if (ticket != null) {
+			throw new TicketAlreadyRunning();
+		}
+	}
+	
+
+	@Override
+	@Transactional
+	public Long saveTicket(ParkingTicket ticket) {
+		ParkingTicket savedTicket = ticketRepository.save(ticket);
+		return savedTicket.getId();
+	}
+
+	@Override
+	@Transactional
+	public void stopTicket(Long ticketId) {
+		ParkingTicket ticket = findById(ticketId);
+		if (!ticket.isStopped()) {
+			ticket.setStopTime(LocalDateTime.now());
+			ticket.setStopped(true);
+			saveTicket(ticket);
+		} else {
+			throw new TicketAlreadyStopped();
```

Comment:
czy to powinien być błąd?

Reply:
W zasadzie nie, można by nie robić nic jeśli juz jest zastopowany